Add Insight note

becdetat opened the pull request:
Fixes #4210

There is a requirement in the ticket that the button text change after being clicked. That has been implemented via a new column in the `wp_wc_admin_note_actions` table - `actioned_label`.

No specific tracks events were added because there is [a comprehensive event that is sent whenever a note action is triggered](https://github.com/woocommerce/woocommerce-admin/blob/master/src/API/NoteActions.php#L118).

### Screenshots

![image](https://user-images.githubusercontent.com/224531/84000801-bbe26e00-a9a8-11ea-8b0e-1fc5ba845e16.png)
![image](https://user-images.githubusercontent.com/224531/84000822-c43aa900-a9a8-11ea-8e6d-9ba93829f617.png)

### Detailed test instructions:

- Your wc-admin installation must be at least 8 days old
- run the wc-admin-daily cron task
- the Insight note should appear in the admin notes
- Click "Yes". The text should change to "Thanks for your feedback" while the page is reloading.

<!--- Note: When displaying information based on sample data, such as SwaggerHub, 
be sure to detail parts affected in Release Notes --->

### Changelog Note:

<!--- Optional: Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance:. If no note is entered, one will be constructed from the title and labels. --->

becdetat committed:
Add Insight note

joshuatf commented on client/header/activity-panel/panels/inbox/action.js:
```diff
@@ -69,7 +87,7 @@ class InboxNoteAction extends Component {
 				onClick={ this.handleActionClick }
 			>
-				{ dismiss ? label : action.label }
+				{ this.getButtonLabel() }
```
Should the actioned text be up one level in the `Card`?  https://github.com/woocommerce/woocommerce-admin/blob/ce68a4a48bac98d9da51ddd9b6c24cfb509c6f25/client/header/activity-panel/panels/inbox/card.js#L248

I think it might be best to display the text instead of the buttons, especially if there were multiple options.  There may also be some accessibility concerns around displaying buttons after actioning.

becdetat replied:
I agree

joshuatf commented on src/Notes/DataStore.php:
```diff
@@ -228,12 +228,13 @@ private function read_actions( &$note ) {
+					'query'          => $action->query,
+					'status'         => $action->status,
+					'primary'        => (bool) $action->is_primary,
+					'actioned_label' => $action->actioned_label,
```
Maybe `actioned_text` if we go with my above suggestion?

joshuatf commented on src/Notes/WC_Admin_Notes_Insight.php:
```diff
@@ -0,0 +1,59 @@
+		// this. Apart from setting the note to actioned a tracks event is
+		// sent in NoteActions.
+		$note->add_action(
+			'affirm-insight',
```
Side note and definitely outside the scope of this PR, but the number of arguments here is getting pretty long and hard to follow.  We may eventually want to look into using a parsable array of arguments to make this more manageable.

joshuatf commented on src/Notes/WC_Admin_Notes_Insight.php:
```diff
@@ -0,0 +1,59 @@
+		// Note that there is no corresponding function called in response to
+		// this. Apart from setting the note to actioned a tracks event is
+		// sent in NoteActions.
+		$note->add_action(
```
Not a blocker for this PR, but curious if we should be adding negative options for insights as well.  I.e., "No, this was not helpful." /cc @pmcpinto

joshuatf commented on src/Notes/WC_Admin_Notes_Insight.php:
```diff
@@ -0,0 +1,59 @@
+	/**
+	 * Name of the note for use in the database.
+	 */
+	const NOTE_NAME = 'wc-admin-insight';
```
@pmcpinto Will more insights be added in the future?  If so, let's rename this to be more specific, otherwise you can ignore this comment @becdetat.

becdetat replied:
Done in 0f9379669af5ed0706605dbe95f9a24969739264

joshuatf reviewed (commented):
Very cool to see the actioned text on this!  Nice work.  Added a couple minor nitpicky comments, but overall testing well.

becdetat committed:
Change actioned_label to actioned_text

becdetat committed:
Move actioned text up a level

becdetat committed:
Fix comment

becdetat commented:
@joshuatf I've moved the actioned text up a level as you suggested, I think it's a good improvement. I'll wait for @pmcpinto 's feedback before moving this back to review required.

pmcpinto commented:
> Not a blocker for this PR, but curious if we should be adding negative options for insights as well. I.e., "No, this was not helpful." /cc @pmcpinto

That can be useful. In an upcoming iteration, we plan to add a pattern that works for any notification when a user clicks "dismiss". But in the meantime, it makes sense to include the "No, this was not helpful." in this specific case.

> @pmcpinto Will more insights be added in the future? 
Yes, the idea is to start adding more insights in the future.

becdetat committed:
Rename insight class

becdetat committed:
Add deny action to insight first sale and fix the button actions

becdetat commented:
I've added the negative option (which was relatively easy) and fixed a little logic error that was stopping the tracks to work (which was surprisingly hard) in 7a77be00. Ready for another review if you wouldn't mind @joshuatf .

becdetat committed:
Merge branch 'master' into add-insight-note

becdetat committed:
fix spacing

joshuatf commented on client/header/activity-panel/panels/inbox/card.js:
```diff
@@ -27,6 +27,7 @@ class InboxNoteCard extends Component {
 		this.state = {
 			isDismissModalOpen: false,
 			dismissType: null,
+			clickedActionText: '',
```
Nice- I really liked the way you handle this here!

What do you think about setting the default to `null` for this value?  Might make the conditions below easier to manage.

becdetat replied:
Sure

becdetat replied:
Done in 3a6052a4

joshuatf commented on src/Notes/WC_Admin_Notes_Insight_First_Sale.php:
```diff
@@ -0,0 +1,67 @@
+			false,
+			__( 'Thanks for your feedback', 'woocommerce-admin' )
+		);
+		$note->add_action(
```
Nice!

joshuatf commented on client/header/activity-panel/panels/inbox/card.js:
```diff
@@ -247,22 +254,34 @@ class InboxNoteCard extends Component {
 	}
 
+	onActionClicked = ( action ) => {
+		this.setState( {
```
Maybe we can return early here if no action text exists:

```js
if ( ! action.actioned_text ) {
	return;
}
```

becdetat replied:
Done in 3a6052a4

joshuatf reviewed (commented):
Nice refactor here, @becdetat!  Left a couple quick comments.

I'm noticing now that the page does a full refresh after clicking an action, which means the actioned text flashes on screen for a short moment before the page refreshes.  Do you see this behavior or is it just my local install?  It may be related to the new `onClick` function.

becdetat commented:
@joshuatf 

> I'm noticing now that the page does a full refresh after clicking an action, which means the actioned text flashes on screen for a short moment before the page refreshes. Do you see this behavior or is it just my local install? It may be related to the new onClick function.

I have the same behaviour but that's just how the href-less actions work - the API call is made which actions the note, and the empty href causes a navigation to the current URL.

becdetat committed:
Tweak conditions in the inbox card

joshuatf reviewed (approved):
Thanks for the all the work on this PR, Bec!

>I have the same behaviour but that's just how the href-less actions work - the API call is made which actions the note, and the empty href causes a navigation to the current URL.

I don't believe that is the intended behavior.  See https://github.com/woocommerce/woocommerce-admin/pull/3039 where async actions were added to notes to avoid page navigation.

But I think this is also broken on master, as seen in the `WC_Admin_Notes_Install_JP_And_WCS_Plugins` note, which no longer works as intended.  So I think this PR is good to go for now and we can tackle those async note action issues in a follow-up.a

jeffstieler commented:
@becdetat as a heads up - I switched the base branch to `main`. You might need to rebase.

becdetat committed:
Merge branch 'main' into add-insight-note

pmcpinto commented:
@becdetat is Yes and No triggering the same event prop?

becdetat commented:
@pmcpinto from what I can see (in `src/API/NoteActions.php` within `trigger_note_action` for the record) it is the same event, but with a unique `action_name` prop (as well as an `action_label` prop) which you should be able to use to differentiate between which action was selected.

pmcpinto commented:
Gotcha. Thanks for the clarification!
